### ISSUE TITLE
OOM with the big stats object

### ISSUE THREAD

**Issue by khanguy00** [FormidableLabs/webpack-stats-plugin#84]
We have a project that the stats object may be larger than 10GB, so using `JSON.stringify()` is not feasible. Is it possible to use https://github.com/Faleij/json-stream-stringify instead of the current approach?

**Comment by ryan-roemer** [comment]
You should be able to limit OOM already with:

1. Use https://github.com/FormidableLabs/webpack-stats-plugin#custom-stats-configuration to ensure you're only passing just the stats fields you need at the webpack level; and then
2. Use https://github.com/FormidableLabs/webpack-stats-plugin#custom-transform-function where you can pass anything you want to `transform()` to serialize including the library you referenced.

Let us know how it goes!

**Comment by khanguy00** [comment]
We actually tried both but didn't help. Because we need the module list with its dependencies (aka `reasons`) to analyze, so the string representation would be huge (more than 16GB), but we can't just keep increasing `max-old-space-size` whenever it's OOM-ed.

I'm considering forking this project and working on `JSONStreamStringify` instead.

**Comment by ryan-roemer** [comment]
If you look at our source, the only use of JSON.stringify is in transform, which it sounds like you already tired to swap out with the stream version. I’m not sure what further changes to this plugin would actually address your issue. What did your webpack configuration look like when you tried to use the json stream library on transform?

**Comment by khanguy00** [comment]
I tried this:

```
new StatsWriterPlugin({
  filename: 'webpack-stats.json',
  stats: {
    timings: false,
    source: false,
    assets: false,
    children: false,
    chunksSort: 'name',
    modules: false,
  },
  transform: (data: any) => {
    const jsonStream = new JsonStreamStringify(data, undefined, 2);
    return new Promise((resolve, reject) => {
      const _buf: string[] = [];
      jsonStream.on('data', (chunk: string) => _buf.push(chunk));
      jsonStream.on('end', () => resolve(_buf.join('')));
      jsonStream.on('error', (err: any) => reject(`error converting stream - ${err}`));
    });
  }
})
```

And then the OOM happened when it pushes too many things to `_buf` array.

I need that because of this:

https://github.com/FormidableLabs/webpack-stats-plugin/blob/a762b43a1e6bf7a8f84bf77c88b1982afadeb1af/lib/stats-writer-plugin.js#L137-L139

As you can see, it needs to push something to `new Buffer()`, so basically another in-memory serialization.

I think we need a way to write a stream to file directly (without in-memory serialization), such as https://nodejs.org/en/knowledge/advanced/streams/how-to-use-fs-create-write-stream/

**Comment by ccortezaguilera** [comment]
Hello I stumbled upon this issue because I have stat object that is also pretty large. I tried the PR that you, @ryan-roemer, created and it fits my use case and works great! Can that PR get merged?

**Comment by ryan-roemer** [comment]
Released in `webpack-stats-plugin@1.1.0`

/cc @ccortezaguilera (thanks for the comment!)
